File: nodejs/patterns/src/structural/composite.js
class FileNode {
  constructor(bytesSize) {
    this.bytesSize = bytesSize;
  }

  size() {
    return this.bytesSize;
  }
}

class FolderNode {
  constructor() {
    this.children = [];
  }

  add(child) {
    this.children.push(child);
  }

  size() {
    return this.children.reduce((total, child) => total + child.size(), 0);
  }
}

function demo() {
  const root = new FolderNode();
  root.add(new FileNode(10));
  const nested = new FolderNode();
  nested.add(new FileNode(5));
  nested.add(new FileNode(15));
  root.add(nested);
  return root.size();
}

module.exports = { FileNode, FolderNode, demo };

File: python/patterns/src/prototype_patterns/structural/composite.py
"""Composite pattern example."""

from __future__ import annotations

from abc import ABC, abstractmethod


class Node(ABC):
    @abstractmethod
    def size(self) -> int:
        raise NotImplementedError


class File(Node):
    def __init__(self, bytes_size: int) -> None:
        self._bytes_size = bytes_size

    def size(self) -> int:
        return self._bytes_size


class Folder(Node):
    def __init__(self) -> None:
        self._children: list[Node] = []

    def add(self, child: Node) -> None:
        self._children.append(child)

    def size(self) -> int:
        return sum(child.size() for child in self._children)


def demo() -> int:
    root = Folder()
    root.add(File(10))
    nested = Folder()
    nested.add(File(5))
    nested.add(File(15))
    root.add(nested)
    return root.size()
